### DISCUSSION TITLE
Applying a set of transformations to all markdown files and creating a new  summary html file.

### DISCUSSION THREAD

**Discussion by vigneshc** [statiqdev/Discussions#60] [Q&A]
I am new to Statiq, I got the based Markdown to html working with Framework. I would like to add the following
* Add a header and footer for all markdown files, based on what is in the front matter.
* Generate a New html that summarizes these markdown files, it doesn't have a corresponding input file.

Todo both it looks like I would have to write `cshtml`. But I am having trouble connecting the cshtml to the .md files. Also, how do I branch and do different a pipeline and do different set of actions for a subset of files. I see that cleanblog does some of this, but couldn't figure out how with reasonable time.

**Comment by daveaglick** [comment]
> Todo both it looks like I would have to write cshtml

That's probably the easiest way right now. Statiq actually supports several [template engines](https://statiq.dev/web/content-and-data/templates/), though not all of them are great at being "outer" layouts. The way Statiq Web is set up out-of-the-box is to run every `.md` file through the Markdown renderer and then through the Razor engine:

`.md` (input file) -> Markdown engine -> Razor engine -> `.html` (output file)

 If there are no `.cshtml` files in the input folder, then the Razor engine is still run but results in nothing changing since the output from the Markdown renderer is plain HTML. The key thing to understand here is that the Razor engine will pick up on _layouts_ though, even for the Markdown files. So the trick is to create a Razor layout and tell the Markdown files to use it when they run through the Razor engine. The [CleanBlog theme](https://github.com/statiqdev/CleanBlog) is a good example of this, but you can do it for your own sites too. The trick is to add a `_Layout.cshtml` file to your `input` folder that contains the overall layout for all your pages:

```
<!DOCTYPE html>
<html lang="en">
<head>
  ...
</head>
<body>
  ...
  <div>@RenderBody()</div>
  ...
</body>
</html>
```

The `@RenderBody()` call is Razor syntax for "render the actual content here".

> Add a header and footer for all markdown files, based on what is in the front matter.

This is where the layout comes in handy. You could add different layouts for different types of documents, but it may be easier to add everything to a single layout for now. The key to understanding this part is that `Document` in a property in the Razor page that contains the current document, and the current document acts like a dictionary with all the [front matter and other metadata](https://statiq.dev/framework/documents/accessing-metadata) you've added.

So to add a footer with values from your front matter, you might write something like this:

```
<!DOCTYPE html>
<html lang="en">
<head>
  ...
</head>
<body>
  ...
  <div>@RenderBody()</div>
  ...
  <div>
    <p>This is the footer</p>
    @{
      string author = Document.GetString("Author");
      if (!string.IsNullOrEmpty(author))
      {
        <p><strong>Author: </strong> @author</p>
      }
    }
  </div>
</body>
</html>
```

Note that the CleanBlog theme already contains some support for this kind of thing and you only need to create a `_post-footer.cshtml` file in your `input` folder to specify a specific footer for blog posts. If nothing else, it might serve as an example of how to create a more complex set of Razor layouts.

> Generate a New html that summarizes these markdown files, it doesn't have a corresponding input file.

This is a perfect example of the [archives](https://statiq.dev/web/content-and-data/archives) feature. You can create a Razor file and specify which other files it should traverse using front matter. Then you use Razor syntax to render lists or other summary data. The documentation contains [several archives examples](https://statiq.dev/web/examples/archives).
